Paper: How the Scientific Python ecosystem helps answer fundamental questions of the Universe

matthewfeickert opened the pull request:
<!-- If you are creating this PR in order to submit a draft of your paper, please name your PR with `Paper: <title>`. An editor will then add a `paper` label and GitHub Actions will be run to check and build your paper. -->

<!-- See the [project readme](https://github.com/scipy-conference/scipy_proceedings#preview-your-paper) for more information. -->

This PR adds the `myst.md` paper source for the SciPy 2024 proceedings for the talk "How the Scientific Python ecosystem helps answer fundamental questions of the Universe".

_Editor_: <!--editor-->Hongsup Shin @hongsupshin  <!--end-editor-->

_Reviewers_: <!--reviewers-list--> 

- Cainã Max Couto da Silva @cmcouto-silva
- Marcus Hill @Marcdh3

<!--end-reviewers-list-->

matthewfeickert committed:
feat: Add draft stage of paper

matthewfeickert commented:
This has been placed in draft mode as I'm still continuing to write. I'm opening the PR now to ensure that it is created before the May 31 deadline.

matthewfeickert commented:
The `curvenote / publish / check` checks are failing as

```
 ❯ content (22/29 tests passed)
    × DOI Exists: Citation does not have DOI: S2I2HEPSP
    × DOI Exists: Citation does not have DOI: PyHEP
    × DOI Exists: Citation does not have DOI: nanobind
    × DOI Exists: Citation does not have DOI: ATL-SOFT-PUB-2021-001
    × DOI Exists: Citation does not have DOI: Dask
    × DOI Exists: Citation does not have DOI: Dask-awkward
    × DOI Exists: Citation does not have DOI: Dask-histogram
```

If I'm citing things that don't have DOIs (like software that doesn't have an archive on Zenodo) is there a way to communicate this to Curvenote to make it clear this is intended?

leewujung commented:
@matthewfeickert you can add citation keys you want to ignore in `myst.yml` under `error_rules`:
```yaml
  error_rules:
    - rule: doi-exists
      severity: ignore
      keys:
        - Atr03
        - terradesert
```

ameyxd commented:
@matthewfeickert can you make this update so we can run checks?

cbcunc commented:
@matthewfeickert Just a reminder that first submissions must be compete by today. Please do what you need to do to get your PR out of the draft state so we can mark it ready for review and assign a reviewer. We assign reviewers to papers all at once and not piecemeal paper by paper.

matthewfeickert commented:
> We assign reviewers to papers all at once and not piecemeal paper by paper.

I don't understand this last comment, but I'll assume that it is not something that affects actionable information on my side. Thanks for the reminder that the deadline is today at 23:59 Pacific. :+1:

matthewfeickert committed:
feat: Add Vangelis's language edits

Co-authored-by: Vangelis Kourlitis <evangelos.kourlitis@cern.ch>

hongsupshin commented:
> > We assign reviewers to papers all at once and not piecemeal paper by paper.
> 
> I don't understand this last comment, but I'll assume that it is not something that affects actionable information on my side. Thanks for the reminder that the deadline is today at 23:59 Pacific. 👍

No worries @matthewfeickert, we assigned two reviewers to your paper :)

fwkoch commented on papers/matthew_feickert/uncovering-higgs.md:
````diff
@@ -0,0 +1,97 @@
+<!-- https://mystmd.org/guide/directives#directive-include -->
+```{include} code/read.py
+:lang: python
+:caption: Using `coffea`, tree structured ROOT files are read with `uproot` from an efficient file cache, and the relevant branches for physics are filtered out into Awkward arrays.
+The operation is scaled out on a Dask cluster for read performance.
````
Directive options are parsed line-by-line - that means the second sentence here will be lost unless you move it to the previous line.

```suggestion
:caption: Using `coffea`, tree structured ROOT files are read with `uproot` from an efficient file cache, and the relevant branches for physics are filtered out into Awkward arrays. The operation is scaled out on a Dask cluster for read performance.
```

matthewfeickert committed:
Place all directive options on common line

Directive options are parsed line-by-line, so if a directive is split across multiple lines then only the first line will be captured.

Co-authored-by: Franklin Koch <franklin@curvenote.com>

Marcdh3 commented:
I believe this paper does a good job of telling the story that complex, large-scale physics research is now feasible in Python due to recent advancements in the scientific Python ecosystem. The background information clearly defined
  the technological gap that needed to be addressed to make this type of research possible. The authors were thorough in indicating how each library addressed the former challenges, and how such code should be used in an
  analysis workflow. Finally, I believe the Higgs Boson example was a good choice because it is a convincing real-world problem that demonstrated the ease at which this research can now be conducted in Python.

Marcdh3 commented on papers/matthew_feickert/introduction.md:
```diff
@@ -0,0 +1,18 @@
+
+The field of high energy physics (HEP) is devoted to the study of the fundamental forces of Nature and their interactions with matter.
+To study the structure of the Universe on the smallest scales requires the highest energy density environments possible &mdash; similar to those of the early Universe.
+These extreme energy density environments are created at the CERN international laboratory, in Geneva, Switzerland, using the Large Hadron Collider (LHC) to collide bunches of billions of protons at a center-of-mass energy of $\sqrt{s} = 13~\mathrm{TeV}$.
```
For clarity, "bunches of" may be removed.

matthewfeickert replied:
We actually do collide "bunches" of protons (that's the terminology we use too) at a time. (c.f. https://home.cern/news/news/accelerators/lhc-racks-records though we don't actually have a great public explainer in just text) It would be less correct to remove this, but if putting "bunches" in quotations makes it more clear we could do that.

cbcunc replied:
Quoting seems reasonable. The citation seems also an additional clarifier.

Marcdh3 commented on papers/matthew_feickert/introduction.md:
```diff
@@ -0,0 +1,18 @@
+Motivated by the broad success of the Scientific Python ecosystem across many domains of science, and the rise of the Scikit-HEP ecosystem of Pythonic tooling for particle physics [@Rodrigues:2020syo;@henry_schreiner-proc-scipy-2022] and community tools produced by the Institute for Research and Innovation in Software for High Energy Physics (IRIS-HEP) [@S2I2HEPSP;@CWPDOC], there has been a broad ...
+The use of dataframes and array programming for data analysis has enhanced the user experience while providing efficient computations without the need of coding optimized low-level routines.
+The ATLAS collaboration has further extended this ecosystem of tooling to include high-level custom Python bindings to the low level `C++` frameworks using `nanobind` [@nanobind].
+Collectively, these tools are modernizing the methods which researchers are engaging data analysis at large scale and providing a nobel end-to-end analysis ecosystem for the ATLAS collaboration.
```
```suggestion
Collectively, these tools are modernizing the methods which researchers are engaging data analysis at large scale and providing a novel end-to-end analysis ecosystem for the ATLAS collaboration.
```

Marcdh3 commented on papers/matthew_feickert/scientific-python-ecosystem.md:
```diff
@@ -0,0 +1,33 @@
+The data structure of each event consists of variable length lists of physics objects (e.g. electrons, collections of tracks from charged objects).
+To study the properties of the physics objects in a statistical manner, a fixed event analysis procedure is repeated over billions of events.
+This has traditionally motivated the use of "event loops" that implicitly construct event-level quantities of interest and leveraged the `C++` compiler to produce efficient iterative code.
+This precedent made it difficult to take advantage of array programming paradigms that are common in Scientific Python given NumPy [@numpy] vector operations.
```
It would be beneficial to the audience to include a sentence explaining what array programming is and why it is useful in this context.

matthewfeickert replied:
I'm hesitant to have this paper have the responsibility of explaining these concepts to the audience, given that we already provide a reference to the NumPy paper [Array programming with NumPy](https://www.nature.com/articles/s41586-020-2649-2). The NumPy paper doesn't make any attempt to further cite the concept of array programming or to elaborate on its definition any further than the first sentence of the paper, I would assume, given how widespread these concepts are in scientific computing. We do also mention in the introduction

> The use of dataframes and array programming for data analysis has enhanced the user experience while providing efficient computations without the need of coding optimized low-level routines.

Marcdh3 commented on papers/matthew_feickert/uncovering-higgs.md:
```diff
@@ -0,0 +1,96 @@
+
+The most famous and revolutionary discovery in particle physics this century is the discovery of the Higgs boson &mdash; the particle corresponding to the quantum field that gives mass to fundamental particles through the Brout-Englert-Higgs mechanism &mdash; by the ATLAS and CMS experimental collaborations in 2012. [@HIGG-2012-27;@CMS-HIG-12-028]
+This discovery work was done using large amounts of customized `C++` software, but in the following decade the state of the PyHEP community has advanced enough that the workflow can now be done using community Python tooling.
+To provide an overview of the tooling and functionality, a high level summary of a simplified analysis workflow of a Higgs "decay" to two intermediate $Z$ bosons that decay to charged leptons $(\ell)$ (i.e. electrons ($e$) and muons ($\mu$)), $H \to Z Z^{*} \to 4 \ell$, on ATLAS open data [@ATLAS-open-data] is summarized in this section.
```
(Optional) Restructure the first paragraph in this section by putting the message of the last sentence first. For example, "In this section we will demonstrate how the scientific Python ecosystem allows for ...  with an example of a simplified analysis workflow of a Higgs Boson decay." My reasoning for this suggestion is because during my first read through the paper I missed the subtle transition into the example; it felt like the paper jumped straight into it after giving background regarding the Higgs Boson phenomenon. So by moving the message of the final sentence up, I think it will stand-out more to the reader.

matthewfeickert replied:
I understand your point, but I don't think we want to restructure that paragraph as this now means awkwardly introducing the Higgs in the middle of another sentence. The focus of the paper is the software, but we still need to motivate it with the science which is the primary driver and so some text needs to be devoted to the physics. Though to try to make the final sentence of the paragraph more clear we've repeated that it is the Pythonic tooling being used. Though if you feel that this still isn't helpful I'm happy to wordsmith this more.

Marcdh3 commented on papers/matthew_feickert/uncovering-higgs.md:
```diff
@@ -0,0 +1,96 @@
+### Cleaning and selecting data
+
+Once the data is in Awkward arrays, additional selections need to be applied before it can be analyzed.
+Only physics object of adequate quality are kept for further analysis and those should reconstruct the topology of interest.
```
```suggestion
Only physics objects of adequate quality are kept for further analysis and those should reconstruct the topology of interest.
```

Marcdh3 commented on papers/matthew_feickert/uncovering-higgs.md:
```diff
@@ -0,0 +1,96 @@
+
+### Feature engineering: The invariant mass
+
+In order to discriminate the events of interest, i.e. candidates of the Higss boson decay, from the vast background which has the same experimental signature, a discriminating feature is constructed.
```
```suggestion
In order to discriminate the events of interest, i.e. candidates of the Higgs boson decay, from the vast background which has the same experimental signature, a discriminating feature is constructed.
```

Marcdh3 commented on papers/matthew_feickert/uncovering-higgs.md:
````diff
@@ -0,0 +1,96 @@
+```{math}
+m = {\frac{\sqrt{E^2 - p(c)^2}}{c^2}}
+```
+where $E$ and $p$ is the total energy and momentum of the particles respectively.
````
```suggestion
where $E$ and $p$ is the total energy and momentum of the particles, respectively.
```

Marcdh3 commented on papers/matthew_feickert/uncovering-higgs.md:
```diff
@@ -0,0 +1,96 @@
+These histograms are then serialized into files with `uproot` and used by the statistical modeling and inference libraries `pyhf` [@pyhf_zenodo;@pyhf_joss] and `cabinetry` [@cabinetry_zenodo] to build binned statistical models and efficiently fit the models to the observed data using vectorized computations and the optimization library `iminuit` [@iminuit_zenodo] for full uncertainties on all mod ...
+The resulting best-fit model parameters &mdash; such as the scale factor on the signal component of the model corresponding to the normalization on the Higgs contributions &mdash; are visualized in @fig:postfit_plot, where good agreement between the model predictions and the data is observed.
+The signal component, clearly visible above the additional "background" components of the model, are Higgs boson events, with an observed count in agreement with theoretical expectations.
+_TODO: Quantify results._
```
Remove this line
```suggestion

```

Marcdh3 reviewed (changes requested):
I believe this paper does a good job of telling the story that complex, large-scale physics research is now feasible in Python due to recent advancements in the scientific Python ecosystem. The background information clearly defined the technological gap that needed to be addressed to make this type of research possible. The authors were thorough in indicating how each library addressed the former challenges, and how such code should be used in an analysis workflow. Finally, I believe the Higgs boson example was a good choice because it is a convincing real-world problem that demonstrated the ease at which this research can now be conducted in Python.

matthewfeickert commented:
Thanks very much for your review @Marcdh3! We will get to work on incorporating your feedback shortly (hopefully before SciPy starts).

cliffckerr reviewed (commented):
I thought this paper was clear and easy to follow. Some very minor comments:
- Should the abstract be updated to refer to the paper instead of the talk?
- You could include an explanation of why ROOT is preferable to other formats (e.g. Arrow?)
- Typo, "a specific calorimeter subsystems"
- The description of the role of "nanobind" probably doesn't need to be repeated between sections.
- Do all computations currently run on CPUs, and if so, is there interest in exploring GPU implementations?

matthewfeickert commented:
Thanks @cliffckerr for the review! I will address these post haste, but am in week 5 of travel for work and teaching at the moment so it probably won't be until next week.

hongsupshin commented:
Hi @matthewfeickert , just a friendly reminder that all initial reviews are in. I highly recommend you start responding to the comments soon since it'd take time for the reviewers to respond to the changes. Remember that the open review period ends on Sep 2, and you will not be able to make any changes to the manuscript after that point. If you have any questions, please let me know!

matthewfeickert committed:
Apply Marcus Hill typo corrections and grammar suggestions

Co-authored-by: Marcus Hill <hill.marcusd@gmail.com>

matthewfeickert committed:
fix: Address Cliff Kerr's revisions and Marcus Hill suggestions

* Add quotes around 'bunches'.
* Reemphasize that the tooling in question is Pythonic.
* Refer to 'paper' instead of 'talk' in the abstract.
* Note ROOT's almost 30 year history of columnar data structures.
* Add 'Awkward Arrays in Python, C++, and Numba' as a reference.
   - c.f. https://inspirehep.net/literature/1776192
* Correct typo of 'subsystems' to 'subsystem'.
* Add citation to IRIS-HEP AGC Zenodo archive.
   - c.f. https://doi.org/10.5281/zenodo.7274936

matthewfeickert commented:
> Some very minor comments:
>
> * Should the abstract be updated to refer to the paper instead of the talk?

Done.

> * You could include an explanation of why ROOT is preferable to other formats (e.g. Arrow?)

I've added a short mention of it providing columnar data structures with good serialization compression for almost 30 years (1997) (Arrow wasn't released until 2016) along with an additional reference. Though the realities are more complex and well outside of the scope of this short paper, I will note that when the field has well over an exabyte of data stored in the same file format which is used for basically everything there's rather large inertia to switch regardless of motivation.

> * Typo, "a specific calorimeter subsystems"

Thanks! @cliffckerr I'm sorry that there were so many typos that you and @Marcdh3 had to catch (and this is even with the `codespell` pre-commit hook on). :grimacing:

> * The description of the role of "nanobind" probably doesn't need to be repeated between sections.

Can you elaborate a bit more on what is repeated? "nanobind" only appears in the main text in the following two sentences:

> The ATLAS collaboration is further extending this ecosystem of tooling to include high-level custom Python bindings to the low level C++ frameworks using nanobind Jakob, 2022.

and

> To expose these C++ libraries to the Pythonic tooling layer, custom Python bindings are written using nanobind for high efficiency, as seen in Figure 1.

The first sentence is introducing the concept of custom Python bindings, and the second is highlighting the fact that `nanobind` in particular was chosen given that the actual bindings it generates are more efficient than alternative binding generation tools. While connected, these ideas are distinct. Do you feel that these concepts need to be emphasized more?

> * Do all computations currently run on CPUs, and if so, is there interest in exploring GPU implementations?

Yes. At the moment the focus has been scaling out these workflows at dedicated "analysis facilities" to achieve required data processing throughput rates for the next iteration of the LHC ([reference presentation](https://indico.cern.ch/event/1369601/contributions/5924000/)). There is ongoing work to be able to utilize hardware acceleration across the tooling ecosystem. This work is still very much in the research stage and at varying stages of maturity across the ecosystem, but [Awkward Array computations work on GPUs](https://awkward-array.org/doc/2.6/user-guide/how-to-math-gpu.html) and the [statistical libraries support hardware acceleration](https://joss.theoj.org/papers/10.21105/joss.02823).

(Note: I rebased this branch off of the current HEAD of the `2024` branch and then did a `git push --force-with-lease`, so my apologies if any comments on the file views got removed from the current view. I've tried to go back and respond to any comments that were added on files.)

cliffckerr commented:
Thanks @matthewfeickert -- my original comment no longer seems so clear to me :joy: It's probably fine as-is. If you do want to revisit the `nanobind` references, I suppose there is a _bit_ of repetition, with each mention of it going into about the same level of detail:
- The ATLAS collaboration is further extending this ecosystem of tooling to include high-level custom Python bindings to the low level `C++` frameworks using `nanobind` [@nanobind].
- To expose these `C++` libraries to the Pythonic tooling layer, custom Python bindings are written using `nanobind` for high efficiency, as seen in @fig:access_layer_diagram.
- The interface takes advantage of `nanobind`'s efficient bindings and zero-copy exchange protocols to achieve viable performance. [@Kourlitis:2890478]
- computed on-the-fly using the `nanobind` Python bindings to the ATLAS `C++` correction tools.

But it's really not an issue, for some reason I thought there was a full paragraph or at least full sentence repeated.

matthewfeickert committed:
fix: Remove information on nanobind performance from Figure 1 caption

* Apply Cliff Kerr's suggestions of avoiding redundant informtion on nanobind.

matthewfeickert commented:
(Sorry for the slow reply @cliffckerr &mdash; been doing a week of international work travel so I'm behind on a bit of everything.)

> * `The interface takes advantage of nanobind's efficient bindings and zero-copy exchange protocols to achieve viable performance. [@Kourlitis:2890478]`
> * `computed on-the-fly using the nanobind Python bindings to the ATLAS C++ correction tools.`

These two come from figure captions, and I generally like them to be as standalone as possible. Though as you do point out, the sentence preceding Figure 1 

> * `To expose these C++ libraries to the Pythonic tooling layer, custom Python bindings are written using nanobind for high efficiency, as seen in @fig:access_layer_diagram.`

makes the mention in the Figure 1 caption

> * `The interface takes advantage of nanobind's efficient bindings and zero-copy exchange protocols to achieve viable performance. [@Kourlitis:2890478]`

a bit redundant as the focus of Figure 1 is the interface, and while the interface decisions enable performance, the performance isn't the focus. So I'm going to remove this line from Figure 1. :+1:

hongsupshin commented:
@cliffckerr @Marcdh3 Hello reviewers, the author @matthewfeickert have responded to your feedback. If you don't have any further comments, could you please approve the PR? Thanks for your time and effort for the review!

hongsupshin commented:
Thanks @Marcdh3 @cliffckerr for reviewing the paper!

cliffckerr commented:
I don't seem to have the option to re-request review, but I approve!

matthewfeickert commented:
> I don't seem to have the option to re-request review, but I approve!

Thanks @cliffckerr! If you'd like to formally hit the button I've triggered review request from you. :) We will take no action to mean "approve" though. :+1:

cliffckerr reviewed (approved):
Looks great, nice work!

matthewfeickert commented on papers/matthew_feickert/mybib.bib:
```diff
@@ -0,0 +1,432 @@
+      reportNumber  = "ATL-COM-SOFT-2024-014",
+      address       = "Geneva",
+      year          = "2024",
+      url           = "https://cds.cern.ch/record/2890478",
```
```suggestion
      url           = "https://indico.cern.ch/event/1330797/contributions/5796636/",
```

@hongsupshin this is very embarrassing, but I realized today while writing a talk that the URL in my reference manager for this poster had an ATLAS internal document server URL for the poster and not the public conference version. :grimacing: Though this is incredibly late would you mind if this got fixed with this suggested change? I didn't notice this before, given that I have access to the internal server and so my cached credentials took me to the document server page.

hongsupshin replied:
Hey Matthew! No worries. Thanks for catching it. We are in the process of merging PRs and getting ready for publishing all the papers. This PR hasn't been merged yet, so I'd say you can commit this change as soon as possible. I will talk to other chairs to freeze the process for this PR in the meantime. Would this work?

matthewfeickert replied:
@hongsupshin yes, and done. :+1: No need to wait as the suggested change (that I just commited) was the only fix needed. Thanks very much!

matthewfeickert committed:
Fixup ATLAS poster URL

* c.f. https://indico.cern.ch/event/1330797/contributions/5796636/

matthewfeickert commented:
Thanks very much to @cliffckerr @Marcdh3 for their hard work as reviewers to help improve the paper &mdash; it is truly appreciated. Thanks also to @hongsupshin and the rest of the Proceedings Committee for the heroic amounts of work that they did to once again serve the SciPy community and provide a wonderful service to the conference. :pray:
